malariagen_data/af1.py

```diff
@@ -0,0 +1,747 @@
+            self.results_cache_set(name=name, params=params, results=results)
+
+        ac = results["ac"]
+        return ac
+
+    def pca(
+        self,
+        region,
+        n_snps,
+        thin_offset=0,
+        sample_sets=None,
+        sample_query=None,
+        site_mask=DEFAULT_SITE_MASK,
+        min_minor_ac=2,
+        max_missing_an=0,
+        n_components=20,
+    ):
+        """Run a principal components analysis (PCA) using biallelic SNPs from
+        the selected genome region and samples.
+
+        Parameters
+        ----------
+        region : str
+            Contig name (e.g., "2RL"), gene name (e.g., "gene-LOC125762289") or
+            genomic region defined with coordinates (e.g.,
+            "2RL:44989425-44998059").
+        n_snps : int
+            The desired number of SNPs to use when running the analysis.
+            SNPs will be evenly thinned to approximately this number.
+        thin_offset : int, optional
+            Starting index for SNP thinning. Change this to repeat the analysis
+            using a different set of SNPs.
+        sample_sets : str or list of str, optional
+            Can be a sample set identifier (e.g., "1229-VO-GH-DADZIE-VMF00095") or a list of
+            sample set identifiers (e.g., ["1240-VO-CD-KOEKEMOER-VMF00099", "1240-VO-MZ-KOEKEMOER-VMF00101"]) or a
+            release identifier (e.g., "1.0") or a list of release identifiers.
+        sample_query : str, optional
+            A pandas query string which will be evaluated against the sample
+            metadata e.g., "country == 'Burkina Faso'".
+        site_mask : {"funestus"}
+            Site filters mask to apply.
+        min_minor_ac : int, optional
+            The minimum minor allele count. SNPs with a minor allele count
+            below this value will be excluded prior to thinning.
+        max_missing_an : int, optional
+            The maximum number of missing allele calls to accept. SNPs with
+            more than this value will be excluded prior to thinning. Set to 0
+            (default) to require no missing calls.
+        n_components : int, optional
+            Number of components to return.
+
+        Returns
+        -------
+        df_pca : pandas.DataFrame
+            A dataframe of sample metadata, with columns "PC1", "PC2", "PC3",
+            etc., added.
+        evr : np.ndarray
+            An array of explained variance ratios, one per component.
+
+        Notes
+        -----
+        This computation may take some time to run, depending on your computing
+        environment. Results of this computation will be cached and re-used if
+        the `results_cache` parameter was set when instantiating the Ag3 class.
+
+        """
+        debug = self._log.debug
+
+        # change this name if you ever change the behaviour of this function, to
+        # invalidate any previously cached data
+        name = PCA_RESULTS_CACHE_NAME
```

Comment:
Perhaps this could be a class variable, rather than a module variable? That would make it easier to move this function to the base class?

Reply:
Agreed, will do. Thanks!